extensions/user-query.md

```diff
@@ -0,0 +1,99 @@
+---
+title: User query
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "Simon Ser"
+    period: "2025"
+    email: "contact@emersion.fr"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the unprefixed `user-query` CAP name. Instead, implementations SHOULD use the `draft/user-query` CAP name to be interoperable with other software implementing a compatible work-in-progress version. The final version of the specification will use unprefixed CAP names.
+
+## Description
+
+This document describes the format of the `user-query` extension. This extension enables several clients of the same user connected to a server or bouncer to tell each other about which user queries (conversations between two users) are currently opened.
+
+Some servers or bouncers remember the channels joined by a user and restore them after connection registration. This can be used by clients to synchronize channels, but doesn't work for user queries. If another client opens or closes a user query, there is no way for other clients to become aware of this action. This extension fills this gap.
+
+The user queries opened by a user MUST NOT be disclosed to other users by the server without an explicit opt-in.
+
+## Implementation
+
+The `user-query` extension uses the `draft/user-query` capability and introduces two new commands, `QUERYUSER` and `UNQUERYUSER`.
+
+The `draft/user-query` capability MAY be negotiated, and affects which messages are sent by the server as specified below.
+
+### `QUERYUSER` command
+
+The `QUERYUSER` command can be sent by both clients and servers.
+
+This command has the following general syntax:
+
+    QUERYUSER <nick>
+
+The `nick` parameter specifies a single nickname.
+
+When sent from a client, this command signals to the server that the user has opened a query with the specified user. The server MUST reply to a successful `QUERYUSER` client set command using a `QUERYUSER` server command, or using an error message.
+
+When sent from a server, the `QUERYUSER` command signals to the client that the user has a query opened with the specified user. This command MAY be sent by the server after connection registration, alongside the `JOIN` burst. If a `PRIVMSG` is sent to the user while no user query is opened, the server automatically opens a user query before dispatching the message.
```

Comment:
> If a `PRIVMSG` is sent to the user while no user query is opened, the server automatically opens a user query before dispatching the message.

I don't think this is desirable in all cases. Users running `/msg ChanServ OP #foobar` don't expect a query to open automatically, and especially not on all their clients.

Reply:
as a reply, but a `PRIVMSG` is sent (by the sending user) to the NickServ user.

This paragraph says "user" for both the recipient and the sender, so it's not clear which one is meant here.

Reply:
I understood it to refer to the recipient but yeah, some clarification would be useful.

For anything other than *Serv it would probably make sense to open the synced query window on send. Maybe it should be a SHOULD and implementations can decide when to do it (e.g. most implementations would want to exempt *Serv)?

Reply:
Yeah I definitely meant that a query is opened when the connected user receives a PRIVMSG. See the last example. I'm not super sure how to phrase it though, the server performs the action to open a query, and from the server PoV the recipient doesn't _receive_ the message, the server merely forwards it if a client is connected.

I don't think opening a query when the current user sends a PRIVMSG is desirable, let alone *Serv special-casing. Clients can just QUERYUSER/UNQUERYUSER when the user opens a query with the client UI or `/query` + `/close`. Clients will not QUERYUSER on `/msg`, as such no user query will be opened on the server side as expected.